### ISSUE TITLE
gRPC Mock Servers with MbDotNet?

### ISSUE THREAD

**Issue by jzabroski** [mattherman/MbDotNet#116]
I'm looking for a way to speed up iterating on tests with a third party that uses gRPC for their public API.  Their API is kind of slow and prone to failing mid-stream.

It would be awesome if MbDotNet could be explained to me in a way I could use it to isolate a couple of response sequences from their servers and replay them as integration test data.

Happy to donate several cups of coffee to you for help.

**Comment by mattherman** [comment]
It looks like there is a gRPC community plugin for Mountebank:
https://github.com/cbrz/mountebank-grpc

Unfortunately, MbDotNet only supports the official protocols (HTTP, HTTPS, TCP, SMTP) and doesn't currently have a way of creating imposters for custom protocols. If you wanted to create imposters for that protocol from .NET code you'd just need to craft the imposter JSON and POST it to Mountebank yourself rather than use MbDotNet.

As far as your question about replaying response sequences it sounds like you might want to look into the proxy features available in Mountebank:
https://mbtest.org/docs/api/proxies

That should allow you to proxy to the public API and capture predicates/responses on an imposter so that future requests are served by Mountebank rather than the actual API.

MbDotNet supports proxy responses, but again, only for the official protocols. If you want to use that feature for gRPC you would need to use that community plugin and create the imposter yourself.


This does give me ideas for the future though. I think with some small changes I may be able to allow people to create their own imposter implementations for custom protocols via inheriting from some of the package types. That would avoid adding those custom protocols to the package while allowing it to be more extensible.

**Comment by mattherman** [comment]
I was exploring what extending MbDotNet to support a custom protocol might look like. I used the Telnet custom protocol provided [here](https://github.com/telekom/mb-netmgmt) as a test since it is very simple. It looks like there is a single request field, `command`, and a single response field, `response`.

This is the implementation of that custom protocol using MbDotNet:
https://github.com/mattherman/mbdotnet-custom-protocol-examples/blob/main/src/MountebankCustomProtocol/TelnetImposter.cs

It relies on a couple changes in the MbDotNet package that I haven't merged yet, but it gives you an idea of what's involved.

You pretty much just create your predicate fields (extending `PredicateFields`), response fields (extending `ResponseFields`), stub (extending `Stub`), and finally your imposter class (extending `Imposter`). The predicate and response field classes need to have their properties tagged with the proper JSON names as defined by the custom protocol itself. The imposter class also needs to pass the appropriate protocol identifier to its base class (in this case `"telnet"`).

_Disclaimer: I haven't tested this quite yet since I didn't have time to get the custom protocol set up, but I'm planning on trying that._

Once I publish the changes I made to support this pattern you would be able to do the same thing with the gRPC custom protocol I linked above.
